### ISSUE TITLE
Weird Behavior When Using `status.hostIP` as datadog hostname in kubernetes

### ISSUE THREAD

**Issue by bangau1** [DataDog/java-dogstatsd-client#47]
Hi,

We have encountered some problem related to networking in kubernetes GKE. This is very weird issue. To give you better understanding about our issue, let me explain the background a little bit.

So, we're using daemonset datadog, to ensure only 1 pod/instance of dd-agent (the name of datadog container/pod) in every node. We expose the `hostPort:8125` for the daemonset
Here is the snippet of our dd-agent daemonset object:
```
ports:
- name: dogstatsdport
  containerPort: 8125
  hostPort: 8125
  protocol: UDP

```

By using the hostPort:8125, other pods in the same node/host can communicate with dd-agent pod via `status.hostIP` (Downward API). Why we prefer this way (using status.hostIP), because the locality and correctness aspect (the pod should communicate with dd-agent in the same node for better latency and correctness of tagging the metrics with the same hostname/node).

Hopefully you have better understanding at this point.

So, let me explain the problem.

We're using java-datadog-statsd-client (https://github.com/DataDog/java-dogstatsd-client), it uses UDP protocol to send metrics to dd-agent statsd server. 
When the dd-agent is up prior to the client pod (the pod who send the metrics), the metrics is sent successfully (indicated by the metrics shown in datadog dashboard).
However, when we restart the dd-agent pod (by killing the pod -> to let the daemonset spawn another dd-agent pod in the same node), the metrics doesn't show up in the datadog dashboard. To solve this, we have to manually restart the client pod, then the metrics is shown.

So, to summarize, in GKE environment, this scenario:
  - dd-agent pod is up
  - client pod is up
  - client pod sending the metrics to dd-agent pod in the same node/host
  - metrics/data is shown at datadog dashboard 
  - dd-agent is restarted (killed, and new pod is spawned to replace the killed ones, by the daemonset controller)
  - the metrics is stopped being shown at datadog dashboard
  - restart the client pod -> metrics is shown successfully.

Note:  restarting the client pod everytime we need to restart dd-agent is not ideal for us, since we have many client pod running owned by various internal teams.

1. I also successfully reproduce the same scenario by using the custom client application (java) (from our office network, my laptop) to send the metrics to dd-agent hostIp:port (our office is allowed to access internal network of where gke cluster residing)
2. However, I can't reproduce the same scenario if I am using a dd-agent outside the gke network/cluster. In this case, i deployed dd-agent in my laptop. The metrics is shown eventhough I restarted the dd-agent service (in my laptop) multiple times.

Do you have any idea why this is happening? java-datadog-statsd client is using `Datagram` in their NonBlockingStatsdClient.java (https://github.com/DataDog/java-dogstatsd-client/blob/master/src/main/java/com/timgroup/statsd/NonBlockingStatsDClient.java). I also tried to logging any error in the NonBlockingStatsDClient.java, but no error found. 
I am not a network expert, but I suspect that when the dd-agent pod is killed, the datagram object on the client pod(udp protocol though) still pointing the same socket of the killed dd-agent one

**Comment by antoinepouille** [comment]
Hi,

Thanks for opening this issue! I just forwarded this case to engineering, we'll get back to you through the ticket you open separately.

Cheers

**Comment by bangau1** [comment]
@antoinepouille any update about this issue?

**Comment by kcollasarundell** [comment]
@antoinepouille Is there any update on this issue?

**Comment by antoinepouille** [comment]
@bangau1 Did you ask for updates from the linked ticket? Could you please give me the matching support ID so that I can check its status?

**Comment by truthbk** [comment]
Hi all! Please take a look at https://github.com/DataDog/dogstatsd-csharp-client/issues/64, and the ongoing discussion there. This is most probably an issue with K8s CNI and not with our clients. Essentially, there's a `conntrack` session being created that _"hijacks"_ the UDP packets that are expected at the dogstatsd `datadog-agent` container. We're still not 100% sure if the problem is with the CNI per-se, or if its a legit race-condition while waiting for the agent container to come back up and the host "thinking" those UDP packets being received are actually for the host (hence the bad `conntrack` session, if that is the case). The weird bit there is it would be surprising if a `conntrack` session would be created to a port we're not listening at... we might have to read/investigate that a little bit.

TLDR; not really an issue on the client side, and not much we can do to that regard.
